### PR TITLE
Add manual current price support on backend

### PR DESCRIPTION
**Pull request by nebolax**
Add manual current price support on backend.

Adds new price oracle for manual current price which can be set by the user. This oracle has the highest priority by default.

Manual current price will be used when showing balances to the user and when saving snapshots. It won't be used for historical price queries (e.g. during pnl report).

### CODE REVIEW

File: rotkehlchen/tests/fixtures/accounting.py [hunk 1]
```diff
@@ -195,6 +196,7 @@ def create_inquirer(
         data_dir=data_directory,
         cryptocompare=cryptocompare,
         coingecko=gecko,
+        manualcurrent=ManualCurrentOracle(),
```
Comment (yabirgb) [inline rotkehlchen/tests/fixtures/accounting.py:199]: It feel strange the way things are initialized here. Maybe can you just move all the initializations to the arguments?

File: rotkehlchen/externalapis/manual_current_price.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            asset=from_asset,
+        )
+        if manual_current_result is None:
+            return Price(ZERO)
+        current_to_asset, current_price = manual_current_result
+        current_to_asset_price = Inquirer().find_price(
+            from_asset=current_to_asset,
+            to_asset=to_asset,
+        )
+        return Price(current_price * current_to_asset_price)
```
Comment (yabirgb) [inline rotkehlchen/externalapis/manual_current_price.py:32]: why is this needed? To return the price in USD? Shouldn't you check that if to_asset == current to_asset skip this call?

File: rotkehlchen/externalapis/manual_current_price.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    def __init__(self) -> None:
+        super().__init__(oracle_name='manual current price oracle')
+
+    def rate_limited_in_last(self, seconds: Optional[int] = None) -> bool:
+        return False
+
+    def query_current_price(self, from_asset: Asset, to_asset: Asset) -> Price:
+        """Searches for a manually specified current price of the `from_asset`. If finds
+        the price, converts it to a price in `to_asset` and returns."""
+        from rotkehlchen.inquirer import Inquirer  # pylint: disable=import-outside-toplevel
```
Comment (yabirgb) [inline rotkehlchen/externalapis/manual_current_price.py:21]: This is ugly :/

File: rotkehlchen/globaldb/handler.py [hunk 4]
```diff
@@ -979,6 +980,70 @@ def add_single_historical_price(entry: HistoricalPrice) -> bool:
...
+    @staticmethod
+    def add_manual_current_price(
+            from_asset: Asset,
+            to_asset: Asset,
+            price: Price,
+    ) -> None:
+        """Adds manual current price and turns previously known manual current price into a
+        historical manual price.
+        May raise:
+        - InputError if some db constraint was hit"""
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:992]: Is only possible with the assets right?

File: rotkehlchen/inquirer.py [hunk 5]
```diff
@@ -262,19 +267,22 @@ def __new__(
...
             coingecko: 'Coingecko' = None,
+            manualcurrent: 'ManualCurrentOracle' = None,
     ) -> 'Inquirer':
         if Inquirer.__instance is not None:
             return Inquirer.__instance
 
         assert data_dir, 'arguments should be given at the first instantiation'
         assert cryptocompare, 'arguments should be given at the first instantiation'
         assert coingecko, 'arguments should be given at the first instantiation'
+        assert manualcurrent, 'arguments should be given at the first instantiatio'
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:278]:
```suggestion
        assert manualcurrent, 'arguments should be given at the first instantiation'
```

**Review by yabirgb — COMMENTED** [review]
2 more things:

1. For an existing user... the current price oracle won't appear right? he would need to add it since the settings are not updated
2. Isn't the test we mentioned about the recursion missing?

File: rotkehlchen/globaldb/handler.py [hunk 6]
```diff
@@ -979,6 +980,70 @@ def add_single_historical_price(entry: HistoricalPrice) -> bool:
...
+                        HistoricalPriceOracle.MANUAL.serialize_for_db(),
+                        HistoricalPriceOracle.MANUAL_CURRENT.serialize_for_db(),
+                        from_asset.identifier,
+                    ),
+                )
+                write_cursor.execute(
+                    'INSERT INTO '
+                    'price_history(from_asset, to_asset, source_type, timestamp, price) '
+                    'VALUES (?, ?, ?, ?, ?)',
+                    (
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/globaldb/handler.py:1007]: No, update also can if there already is a manual price for the same pair at the same timestamp

File: rotkehlchen/globaldb/handler.py [hunk 7]
```diff
@@ -979,6 +980,70 @@ def add_single_historical_price(entry: HistoricalPrice) -> bool:
...
+                        str(price),
+                    ),
+                )
+        except sqlite3.IntegrityError as e:
+            raise InputError(f'Failed to add manual current price due to: {str(e)}') from e
+
+    @staticmethod
+    def get_manual_current_price(asset: Asset) -> Optional[Tuple[Asset, Price]]:
+        """Reads manual current price of an asset. If no price is found returns None."""
+        with GlobalDBHandler().conn.read_ctx() as read_cursor:
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/globaldb/handler.py:1021]: Why? `price_history` has foreign key relations with `assets` table

File: rotkehlchen/tests/api/test_current_assets_price.py [hunk 8]
```diff
@@ -71,3 +80,196 @@ def test_get_current_assets_price_in_btc(rotkehlchen_api_server):
...
+        json={
+            'from_asset': A_USD.identifier,
+            'to_asset': A_EUR.identifier,
+            'price': '23',
+        },
+    )
+    assert_proper_response(response)
+    assert GlobalDBHandler().get_manual_current_price(A_USD) == (A_EUR, Price(FVal(23)))
+
+    Inquirer()._cached_current_price = {}
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/tests/api/test_current_assets_price.py:136]: > Because the price for usd was queried before?

Yes. Added a comment

File: rotkehlchen/tests/api/test_current_assets_price.py [hunk 9]
```diff
@@ -71,3 +80,196 @@ def test_get_current_assets_price_in_btc(rotkehlchen_api_server):
...
+        Inquirer()._cached_current_price = {}
+
+        response = requests.get(
+            api_url_for(
+                rotkehlchen_api_server,
+                'allbalancesresource',
+            ), json={
+                'ignore_cache': True,
+            },
+        )
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/tests/api/test_current_assets_price.py:272]: Yes, it is possible to query inquirer only. But this is simply another testing approach, right? And I just chose to go with testing the entire endpoint behaviour.

File: rotkehlchen/tests/fixtures/accounting.py (hunk 1 above)
Comment (nebolax) [reply rotkehlchen/tests/fixtures/accounting.py:199]: Ok

File: rotkehlchen/externalapis/manual_current_price.py (hunk 2 above)
Comment (nebolax) [reply rotkehlchen/externalapis/manual_current_price.py:32]: I don't think that this check is needed here since it happens in `Inquirer().find_price()`

File: rotkehlchen/externalapis/manual_current_price.py (hunk 3 above)
Comment (nebolax) [reply rotkehlchen/externalapis/manual_current_price.py:21]: Yes yes, I know. Originally I didn't see how to move it out of here, but now find a way and fixed. Thanks for the comment!

Comment (nebolax) [reply rotkehlchen/externalapis/manual_current_price.py:21]: Yes yes, I know. Originally I didn't see how to move it out of here, but now find a way and fixed. Thanks for the comment!

File: rotkehlchen/globaldb/handler.py (hunk 4 above)
Comment (nebolax) [reply rotkehlchen/globaldb/handler.py:992]: Also if unique constraint will be hit during turning previous manual current price entry into a historical manual price

File: rotkehlchen/globaldb/handler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/globaldb/handler.py:1007]: So maybe split both of them so we can later figure out the error faster?

File: rotkehlchen/tests/api/test_current_assets_price.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/tests/api/test_current_assets_price.py:272]: Isn't the balances resource tested already? If not you are just wasting time setting up everything instead of testing the functionality you want

File: rotkehlchen/globaldb/handler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/globaldb/handler.py:1013]: makes sense

File: rotkehlchen/tests/api/test_current_assets_price.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/tests/api/test_current_assets_price.py:345]: Adjusted tests locations a bit

File: rotkehlchen/globaldb/handler.py [hunk 10]
```diff
@@ -984,6 +986,86 @@ def add_single_historical_price(entry: HistoricalPrice) -> bool:
...
+                    ),
+                )
+            except sqlite3.IntegrityError as e:
+                # if we got an error, there has to be a corresponding entry in the db
+                with GlobalDBHandler().conn.read_ctx() as cursor:
+                    timestamp = cursor.execute(
+                        'SELECT timestamp FROM price_history WHERE source_type=? AND from_asset=?',
+                        (
+                            HistoricalPriceOracle.MANUAL_CURRENT.serialize_for_db(),
+                            from_asset.identifier,
```
Comment (yabirgb) [inline rotkehlchen/globaldb/handler.py:1016]: We do an extra query on case of error to show useful information to the user about how to fix the error. Maybe put that in the code

File: rotkehlchen/inquirer.py [hunk 11]
```diff
@@ -333,6 +341,10 @@ def get_cached_current_price_entry(cache_key: Tuple[Asset, Asset]) -> Optional[C
 
         return cache
 
+    @staticmethod
+    def remove_cached_current_price_entry(cache_key: Tuple[Asset, Asset]) -> None:
+        Inquirer()._cached_current_price.pop(cache_key, None)
```
Comment (yabirgb) [inline rotkehlchen/inquirer.py:346]: The default value is not needed here

File: rotkehlchen/tests/api/test_current_assets_price.py [hunk 12]
```diff
@@ -71,3 +81,226 @@ def test_get_current_assets_price_in_btc(rotkehlchen_api_server):
...
+        contained_in_msg='Not found manual current price to delete',
+        status_code=HTTPStatus.CONFLICT,
+    )
+    # usd manual current price should have not been touched
+    assert GlobalDBHandler().get_manual_current_price(A_USD) == (A_EUR, Price(FVal(25)))
+
+
+@pytest.mark.parametrize('should_mock_current_price_queries', [False])
+def test_manual_current_prices_loop(inquirer):
+    GlobalDBHandler().add_manual_current_price(
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_current_assets_price.py:267]: Having a description about what this test is for would be nice to have since we will forget in a few weeks

File: rotkehlchen/tests/api/test_current_assets_price.py [hunk 13]
```diff
@@ -71,3 +81,226 @@ def test_get_current_assets_price_in_btc(rotkehlchen_api_server):
...
+    assert len(warnings) == 3
+    assert 'from BTC(Bitcoin) to EUR(Euro) since your manual current' in warnings[0]
+    assert 'from USD(United States Dollar) to EUR(Euro) since your manual current' in warnings[1]
+    assert 'from ETH(Ethereum) to EUR(Euro) since your manual current' in warnings[2]
+
+
+@pytest.mark.parametrize('ignore_mocked_prices_for', ['ETH'])
+def test_inquirer_oracles_affect_manual_price(inquirer):
+    """Checks that change of oracles order affects manual current price usage."""
+
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_current_assets_price.py:294]: Remove the space

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 14]
```diff
@@ -913,6 +920,13 @@ def try_insert_mapping(cur):
         token = cursor.fetchone()[0]
         assert token == 'eip155:1/erc20:0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e'
 
+    # Check that oracles were upgraded properly
+    with db_v34.conn.read_ctx() as cursor:
```
Comment (yabirgb) [inline rotkehlchen/tests/db/test_db_upgrades.py:924]:
```suggestion
    # Check that oracles were upgraded properly
    with db_v35.conn.read_ctx() as cursor:
```

but you can reuse the one before. Same for the other read you made

File: rotkehlchen/api/rest.py [hunk 15]
```diff
@@ -3801,24 +3801,50 @@ def add_manual_current_price(
...
+            return self._api_query_for_eth_module(
+                async_query=False,
+                module_name='nfts',
+                method='delete_price_for_nft',
+                query_specific_balances_before=None,
+                asset=asset,
+            )
+        try:
+            GlobalDBHandler().delete_manual_current_price(asset=asset)
+            return api_response(result=OK_RESULT)
```
Comment (yabirgb) [inline rotkehlchen/api/rest.py:3842]: I believe our style is writing this return out of the try/except statment

File: rotkehlchen/api/rest.py [hunk 16]
```diff
@@ -3801,24 +3801,50 @@ def add_manual_current_price(
...
+                price=price,
+            )
+        try:
+            GlobalDBHandler().add_manual_current_price(
+                from_asset=from_asset,
+                to_asset=to_asset,
+                price=price,
+            )
+            Inquirer().remove_cached_current_price_entry(cache_key=(from_asset, to_asset))
+            return api_response(result=OK_RESULT)
```
Comment (yabirgb) [inline rotkehlchen/api/rest.py:3821]: Our style I believe is moving this return out of the try/except statement

**Review by yabirgb — COMMENTED** [review]
Mostly styling but one important thing

File: rotkehlchen/externalapis/manual_current_price.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+                from_asset=current_to_asset,
+                to_asset=to_asset,
+            )
+            return Price(current_price * current_to_asset_price)
+        except RecursionError:
+            self.msg_aggregator.add_warning(
+                f'Was not able to find price from {str(from_asset)} to {str(to_asset)} since your '
+                f'manual current prices form a loop. For now, other oracles will be used.',
+            )
+            return Price(ZERO)
```
Comment (LefterisJP) [inline rotkehlchen/externalapis/manual_current_price.py:41]: 1. Succesful return at the end.
2. Add a comment explaining when this may happen during RecursionError. We discusses when but we will forget it.

```suggestion
        except RecursionError:
            self.msg_aggregator.add_warning(
                f'Was not able to find price from {str(from_asset)} to {str(to_asset)} since your '
                f'manual current prices form a loop. For now, other oracles will be used.',
            )
            return Price(ZERO)

        return Price(current_price * current_to_asset_price)
```

File: rotkehlchen/globaldb/handler.py [hunk 18]
```diff
@@ -984,6 +986,87 @@ def add_single_historical_price(entry: HistoricalPrice) -> bool:
...
+    @staticmethod
+    def add_manual_current_price(
+            from_asset: Asset,
+            to_asset: Asset,
+            price: Price,
+    ) -> None:
+        """Adds manual current price and turns previously known manual current price into a
+        historical manual price.
+        May raise:
+        - InputError if some db constraint was hit. Probably means manual price duplication."""
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/handler.py:998]:
```suggestion
        - InputError if some db constraint was hit. Probably means manual price duplication.
        """
```

File: rotkehlchen/globaldb/handler.py [hunk 19]
```diff
@@ -984,6 +986,87 @@ def add_single_historical_price(entry: HistoricalPrice) -> bool:
 
         return True
 
+    @staticmethod
+    def add_manual_current_price(
+            from_asset: Asset,
+            to_asset: Asset,
+            price: Price,
+    ) -> None:
+        """Adds manual current price and turns previously known manual current price into a
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/handler.py:995]:
```suggestion
        """
        Adds manual current price and turns previously known manual current price into a
```

Here and below kind of a nitpick for docstrings

File: rotkehlchen/globaldb/schema.py [hunk 20]
```diff
@@ -190,6 +190,8 @@
 INSERT OR IGNORE INTO price_history_source_types(type, seq) VALUES ('C', 3);
 /* XRATESCOM */
 INSERT OR IGNORE INTO price_history_source_types(type, seq) VALUES ('D', 4);
+/* MANUAL_CURRENT */
+INSERT OR IGNORE INTO price_history_source_types(type, seq) VALUES ('E', 5);
```
Comment (LefterisJP) [inline rotkehlchen/globaldb/schema.py:194]: This shoudl also go to the global DB upgrade

**Review by LefterisJP — COMMENTED** [review]
Some comments

File: rotkehlchen/api/v1/resources.py [hunk 21]
```diff
@@ -2140,7 +2140,7 @@ def patch(self, asset: Asset) -> Response:
 
 class CurrentAssetsPriceResource(BaseMethodView):
 
-    put_schema = ManualPriceSchema
+    put_schema = ManualPriceSchema()
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/api/v1/resources.py:2143]: I think it works due to this:

https://github.com/marshmallow-code/webargs/blob/dd9520d9ab1b1e00d1a19564d5c78a9fba27c8a9/src/webargs/core.py#L301

File: rotkehlchen/externalapis/manual_current_price.py [hunk 22]
```diff
@@ -0,0 +1,41 @@
+from typing import Optional
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/externalapis/manual_current_price.py:1]: Put them in one file and adjusted the file's name to `manual_price_oracles.py`

File: rotkehlchen/globaldb/schema.py (hunk 20 above)
Comment (nebolax) [reply rotkehlchen/globaldb/schema.py:194]: Oh yes, indeed

**Review by LefterisJP — COMMENTED** [review]
Also make sure the 3 failed tests in the backend pass locally

**Review by LefterisJP — APPROVED** [review]
LGTM

**Review by LefterisJP — APPROVED** [review]
LGTM. Only bisq failed!
